File: backend/handlers/websocket.go
package handlers

import (
	"encoding/json"
	"log"
	"net/http"
	"time"

	"chat-app/room"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

type WSHandler struct {
	Manager *room.Manager
}

func NewWSHandler(manager *room.Manager) *WSHandler {
	return &WSHandler{Manager: manager}
}

type Message struct {
	Type      string `json:"type"`
	Name      string `json:"name"`
	Content   string `json:"content"`
	Timestamp string `json:"timestamp"`
}

func (h *WSHandler) HandleWebSocket(c *gin.Context) {
	token := c.Param("token")
	userName := c.Query("name")
	userIP := c.ClientIP()

	if userName == "" {
		userName = "Anonymous"
	}

	// Claim the invite (single-use)
	r, ok := h.Manager.ClaimInvite(token, userIP)
	if !ok || r == nil {
		c.JSON(http.StatusForbidden, gin.H{"error": "invite already used or invalid"})
		return
	}

	conn, err := upgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Printf("WebSocket upgrade error: %v", err)
		return
	}

	client := &room.Client{
		Conn: conn,
		Name: userName,
	}

	r.AddClient(client)

	joinMsg := Message{
		Type:      "system",
		Name:      userName,
		Content:   userName + " joined the chat",
		Timestamp: time.Now().Format("15:04"),
	}
	broadcastMessage(r, joinMsg)

	defer func() {
		r.RemoveClient(client)
		conn.Close()

		leaveMsg := Message{
			Type:      "system",
			Name:      userName,
			Content:   userName + " left the chat",
			Timestamp: time.Now().Format("15:04"),
		}
		broadcastMessage(r, leaveMsg)
	}()

	for {
		_, msgBytes, err := conn.ReadMessage()
		if err != nil {
			break
		}

		var incoming struct {
			Content string `json:"content"`
		}
		if err := json.Unmarshal(msgBytes, &incoming); err != nil {
			continue
		}

		if incoming.Content == "" {
			continue
		}

		chatMsg := Message{
			Type:      "chat",
			Name:      userName,
			Content:   incoming.Content,
			Timestamp: time.Now().Format("15:04"),
		}
		broadcastMessage(r, chatMsg)
	}
}

func broadcastMessage(r *room.Room, msg Message) {
	data, err := json.Marshal(msg)
	if err != nil {
		return
	}
	r.Broadcast(data)
}

File: backend-node/handlers/websocket.js
const { censorMessage } = require('../utils/filter');

class WSHandler {
  constructor(manager) {
    this.manager = manager;
  }

  handleConnection = (ws, req) => {
    const url = new URL(req.url, `http://${req.headers.host}`);
    const pathParts = url.pathname.split('/');
    const token = pathParts[2];
    const userName = url.searchParams.get('name') || 'Anonymous';

    const { room, success } = this.manager.claimInvite(token);
    if (!success || !room) {
      ws.close(1008, 'Invalid or expired invite');
      return;
    }

    const clientId = Date.now() + Math.random();
    const client = { id: clientId, ws, name: userName, room, role: 'chat' };
    room.addClient(client);

    // Send current pin if exists
    if (room.pinnedMessage) {
      ws.send(JSON.stringify({ type: 'pin', pinnedMessage: room.pinnedMessage }));
    }

    // Notify admin stream about new user
    room.broadcastStatus();

    ws.on('message', (data) => {
      try {
        const incoming = JSON.parse(data.toString());
        if (!incoming.content || incoming.content.trim() === '') return;

        const chatMsg = {
          type: 'chat',
          name: userName,
          content: censorMessage(incoming.content.trim()),
          timestamp: new Date().toLocaleTimeString('en-US', {
            hour12: false,
            hour: '2-digit',
            minute: '2-digit'
          })
        };

        const stored = room.storeMessage(chatMsg);
        room.broadcast(stored);
        room.broadcastStatus();
      } catch (error) {
        console.error('Invalid message format:', error);
      }
    });

    ws.on('close', () => {
      room.removeClient(clientId);
      room.broadcastStatus();

      const leaveMsg = {
        type: 'system',
        content: `${userName} left the chat`,
        timestamp: new Date().toLocaleTimeString('en-US', {
          hour12: false,
          hour: '2-digit',
          minute: '2-digit'
        })
      };
      room.broadcast(leaveMsg);
    });

    ws.on('error', (error) => {
      console.error('WebSocket error:', error);
      room.removeClient(clientId);
      room.broadcastStatus();
    });
  };
}

class StreamHandler {
  constructor(manager) {
    this.manager = manager;
  }

  handleConnection = (ws, req) => {
    const url = new URL(req.url, `http://${req.headers.host}`);
    const pathParts = url.pathname.split('/');
    const roomId = pathParts[2];

    const room = this.manager.getRoom(roomId);
    if (!room) {
      ws.close(1008, 'Room not found');
      return;
    }

    const clientId = Date.now() + Math.random();
    const client = { id: clientId, ws, name: '__stream__', room, role: 'stream' };
    room.addClient(client);

    // Send history, current status and pin to new stream/admin viewer
    if (room.recentMessages.length > 0) {
      ws.send(JSON.stringify({ type: 'history', messages: room.recentMessages }));
    }
    ws.send(JSON.stringify({
      type: 'status',
      userCount: room.getUsers().length,
      users: room.getUsers(),
      messageCount: room._msgCounter
    }));
    if (room.pinnedMessage) {
      ws.send(JSON.stringify({ type: 'pin', pinnedMessage: room.pinnedMessage }));
    }

    ws.on('message', () => { /* stream is read-only */ });

    ws.on('close', () => {
      room.removeClient(clientId);
    });

    ws.on('error', (error) => {
      console.error('Stream WebSocket error:', error);
      room.removeClient(clientId);
    });
  };
}

module.exports = { WSHandler, StreamHandler };
